java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java

```diff
@@ -240,6 +250,22 @@ private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol metho
     return currentSubstitution;
   }
 
+  private static JavaType leastUpperBound(List<JavaType> remainingArgTypes) {
+    return (JavaType) Types.leastUpperBound(typeListToTypeWithWrapperSet(remainingArgTypes));
+  }
+
+  private static Set<Type> typeListToTypeWithWrapperSet(List<JavaType> remainingArgTypes) {
+    Set<Type> results = new HashSet<>();
+    for (Type type : remainingArgTypes) {
```

Comment:
(Very) minor optimisation but you should construct the set and iterate on it,  replacing primitives, that would save some operations.